.github/workflows/ci.yml

```diff
@@ -0,0 +1,445 @@
+            --no-maintainer-edit || true
+
+          printf "Pipeline Attempt on $GITHUB_RUN_ID for $GITHUB_SHA\n\n$GITHUB_SERVER_URL/$GITHUB_REPOSITORY/actions/runs/$GITHUB_RUN_ID" \
+          | gh pr comment staging \
+            --body-file - \
+            --repo "$GH_PROJECT_PATH"
+
+  build-dist:
+    name: "Build / Dist"
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/matrixai/github-runner
+
+    needs:
+      - check-lint
+      - check-dry
+
+    if: github.ref == 'refs/heads/staging'
+
+    steps:
+      - uses: actions/checkout@v4
+      - run: entrypoint
+
+      - name: Build dist
+        run: |
+          nix develop .#ci --command bash -c $'
+            npm run build --verbose
+          '
+
+      - uses: actions/upload-artifact@v4
+        with:
+          name: build-dist_archive
+          path: ./dist
+
+  build-prerelease:
+    name: "Build / Pre-release"
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/matrixai/github-runner
+
+    needs:
+      - build-dist
+
+    if: github.ref == 'refs/heads/staging'
+
+    steps:
+      - uses: actions/checkout@v4
+      - run: entrypoint
+
+      - name: Build dist
+        run: |
+          echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > ./.npmrc
+          echo 'Publishing library prerelease'
+
+          nix develop .#ci --command bash -c $'
+            npm publish --tag prerelease --access public
+          '
+
+          rm -f ./.npmrc
+
+  integration-builds:
+    name: "Integration / Builds"
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/matrixai/github-runner
+
+    needs:
+      - build-pull
+      - build-dist
+
+    if: github.ref == 'refs/heads/staging'
+
+    steps:
+      - uses: actions/checkout@v4
+      - run: entrypoint
+
+      - name: Build targets
+        run: |
+          mkdir -p ./builds
+
+          build-application="$(nix build \
+            --max-jobs "$(nproc)" \
+            --cores "$(nproc)" \
+            --print-out-paths \
+            .# \
+          )"
+
+          nix-store --export $( \
+            nix-store --query --requisites "$build-application" \
+          ) | gzip > ./builds/polykey-cli.closure.gz
+
+          builds="$(nix build \
+            --max-jobs "$(nproc)" \
+            --cores "$(nproc)" \
+            --print-out-paths \
+            --print-build-logs \
+            .# \
+            .#docker \
+            .#packages.x86_64-linux.executable \
+          )"
+
+          cp -r $(echo $builds | tr '\n' ' ') ./builds/
+
+          nix develop .#ci --command bash -c $'
+            npm run build
+          '
+
+      - uses: actions/upload-artifact@v4
+        with:
+          name: builds
+          path: ./builds
+
+  integration-deployment:
+    name: "Integration / Deployment"
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/matrixai/github-runner
+
+    concurrency:
+      group: integration-deployment
+      cancel-in-progress: true
```

Comment:
Why is this needed here? Why do we want to cancel in-progress deployments if a new one is triggered?

Reply:
Concurrency groups were previously used in the `.gitlab-ci.yml` file in the form of resource groups, so this is being mirrored here using concurrency groups instead. This is intentional.